### PR TITLE
SAP port to baml_language

### PR DESCRIPTION
**Pull request by 2kai2kai2**
- Porting jsonish from engine to baml_language
- Porting SAP from engine to baml_language
- Ensuring regression tests succeed
- Using new SAP model with new attributes and new streaming semantics representation

## Summary by CodeRabbit

* **New Features**
  * New SAP/BAML parsing and public parse entrypoint supporting Markdown/code blocks, multi-JSON and fixer fallbacks.
  * Metadata-rich, serializable BAML value model with streaming/partial completion tracking and UI-friendly error/explanations.
  * Accent- and punctuation-tolerant string matching with best-match scoring for enums/aliases.

* **Refactor**
  * Unified, extensible deserialization/coercion framework for literals, primitives, arrays, maps, classes, unions, and stream states.

* **Tests**
  * Large, comprehensive test suite covering parsing, coercion, accents, edge cases, and streaming scenarios.

### CODE REVIEW

File: baml_language/crates/bex_sap/src/helpers/mod.rs [hunk 1]
```diff
@@ -0,0 +1,391 @@
+// pub mod common;
+// use std::{collections::HashSet, path::PathBuf};
+
+// use anyhow::Result;
+// // use baml_types::{
```
Comment (aaronvg) [inline baml_language/crates/bex_sap/src/helpers/mod.rs]: we can delete this code if unused

File: baml_language/crates/baml_type/src/lib.rs [hunk 2]
```diff
@@ -48,6 +47,24 @@ impl TypeName {
...
+        let name = Name::new(*segments.last().expect("path must be non-empty"));
+        let module_path = segments[..segments.len() - 1]
+            .iter()
+            .map(|s| Name::new(*s))
+            .collect();
+        Self {
+            display_name: Name::new(path),
+            name,
+            module_path,
+        }
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/baml_type/src/lib.rs:66]: We only run this on compiler output so this will never happen

File: baml_language/crates/bex_sap/src/jsonish/parser/fixing_parser.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                            .completed_values
+                            .into_iter()
+                            .map(|f| {
+                                let _completion_state = *f.1.completion_state();
+                                Value::FixedJson(f.1.into(), f.2)
+                            })
+                            .collect(),
+                        CompletionState::Incomplete, // TODO: Is it complete?
+                    ),
+                    vec![Fixes::InferredArray],
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/jsonish/parser/fixing_parser.rs:73]: These are pre-existing issues in the jsonish parser that was ported. Leaving for now.

File: baml_language/crates/bex_sap/src/jsonish/parser/dedent.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            }
+        })
+        .skip_while(|&l| l.is_empty())
+        .collect::<Vec<_>>()
+        .join("\n");
+
+    if result.ends_with('\n') && !s.ends_with('\n') {
+        let new_len = result.len() - 1;
+        result.truncate(new_len);
+    }
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/jsonish/parser/dedent.rs:63]: These are pre-existing issues in the jsonish parser that was ported (it used a dedent crate which has this behavior). Leaving for now.

File: baml_language/crates/bex_sap/src/jsonish/value.rs [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    pub fn completion_state(&self) -> &CompletionState {
+        match self {
+            Value::String(_, s) => s,
+            Value::Number(_, s) => s,
+            Value::Boolean(_) => &CompletionState::Complete,
+            Value::Null => &CompletionState::Complete,
+            Value::Object(_, s) => s,
+            Value::Array(_, s) => s,
+            Value::Markdown(_, _, s) => s,
+            Value::FixedJson(v, _) => v.completion_state(),
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/jsonish/value.rs:168]: These are pre-existing issues in the jsonish parser that was ported. Leaving for now.

File: baml_language/crates/bex_sap/src/jsonish/parser/fixing_parser/json_collection.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    pub fn completion_state(&self) -> CompletionState {
+        match *self {
+            JsonCollection::Object(_, _, s) => s,
+            JsonCollection::Array(_, s) => s,
+            JsonCollection::QuotedString(_, s) => s,
+            JsonCollection::SingleQuotedString(_, s) => s,
+            JsonCollection::TripleBacktickString {
+                content: (_, s), ..
+            } => s, // TODO: correct?
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/jsonish/parser/fixing_parser/json_collection.rs:60]: These are pre-existing issues in the jsonish parser that was ported. Leaving for now.

File: baml_language/crates/bex_sap/src/deserializer/coercer/coerce_union.rs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                }
+            }
+        }
+
+        let best = array_helper::pick_best(
+            ctx,
+            TyWithMeta::new(TyResolvedRef::Union(target.ty), target.meta),
+            variants,
+        );
+        best.map(|v| v.with_flags(add_flags)).map(Some)
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/deserializer/coercer/coerce_union.rs]: pdick_best will now handle this behavior

[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/deserializer/coercer/coerce_union.rs]: *pick_best

File: baml_language/crates/bex_sap/src/tests/test_international.rs [hunk 8]
```diff
@@ -0,0 +1,1412 @@
...
+        enum Priority {
+            HIGH @alias("élevé"),
+            MEDIUM @alias("médium"),
+            LOW @alias("baixo"),
+            URGENT @alias("紧急"),
+            NORMAL @alias("عادي")
+        }
+    },
+    // since "médium" -> "MEDIUM"
+    ["MEDIUM"]
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/tests/test_international.rs:418]: This is not inconsistent.

File: baml_language/crates/baml_compiler_emit/src/lib.rs [hunk 9]
```diff
@@ -747,7 +773,7 @@ pub fn compile_files(
             {
                 program
                     .recursive_type_alias_defs
-                    .insert(alias_name.to_string(), target_ty);
+                    .insert(baml_type::TypeName::local(alias_name.clone()), target_ty);
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/baml_compiler_emit/src/lib.rs:776]: There is a major rework in progress for the compiler code so this code will likely be updated when that is ready. Fixing now would be a waste of time.

File: baml_language/crates/bex_sap/src/sap_model/convert.rs [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+pub fn convert_ty_attrs(
+    attrs: &baml_type::TyAttr,
+) -> Result<TypeAnnotations<'_, TypeName>, ConvertError<'_>> {
+    let Some(attrs) = &attrs.0 else {
+        return Ok(TypeAnnotations::default());
+    };
+    Ok(TypeAnnotations {
+        in_progress: Some(convert_attr_literal(&attrs.sap_in_progress)?),
+        asserts: Vec::new(), // TODO: assertions
+    })
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/sap_model/convert.rs:154]: This relies on external code that is currently in flux. Will add once it's stable.

File: baml_language/crates/bex_sap/src/jsonish/parser/multi_json_parser.rs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                        json_str,
+                        options.next_from_mode(super::ParsingMode::AllJsonObjects),
+                        false,
+                    ) {
+                        Ok(json) => json_objects.push(json),
+                        Err(_e) => {
+                            // Ignore errors
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/jsonish/parser/multi_json_parser.rs:48]: This behavior is pre-existing from the ported code. While we may change it in the future, we are currently trying to maintain consistency with the original behavior.

File: baml_language/crates/bex_sap/src/deserializer/coercer/coerce_class.rs [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+            //     };
+            //     let field_meta = DeserializerMeta::new(ty);
+            //     ValueWithFlags::new(field_value, field_meta)
+            //         .with_flag(Flag::OptionalFieldError(name.clone(), e))
+            // }
+            // // Required field with error
+            Some(Err(e)) => {
+                err_unparsed.push((name, e));
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/deserializer/coercer/coerce_class.rs:488]: ignore for now

File: baml_language/crates/bex_sap/src/sap_model/from_literal.rs [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+                        .from_literal(&AttrLiteral::String(key.clone()), ctx)?;
+                    let key =
+                        match key {
+                            BamlValue::String(s) => s.value,
+                            BamlValue::Enum(e) => Cow::Borrowed(e.value), // uses variant name, not aliases
+                            _ => return Err(ctx.error_internal(
+                                "map literal keys must be a string-like type (string, string literal, or enum)",
+                            )),
+                        };
+                    let value = self.value.ty.from_literal(value, ctx)?;
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply baml_language/crates/bex_sap/src/sap_model/from_literal.rs]: leave this for now
